src/v/archival/ntp_archiver_service.cc

```diff
@@ -962,20 +959,29 @@ ss::future<cloud_storage::upload_result> ntp_archiver::do_upload_segment(
         }
     };
 
-    co_return co_await _remote.upload_segment(
-      get_bucket_name(),
-      path,
-      candidate.content_length,
-      std::move(reset_func),
-      fib,
-      lazy_abort_source,
-      _segment_tags);
-}
-
-ss::future<> ntp_archiver::stream_reference::maybe_close_ref() {
-    if (stream_ref.has_value()) {
-        co_await stream_ref->get().close();
+    auto response = cloud_storage::upload_result::success;
+    try {
+        response = co_await _remote.upload_segment(
+          get_bucket_name(),
+          path,
+          candidate.content_length,
+          std::move(reset_func),
+          fib,
+          lazy_abort_source,
+          _segment_tags);
+    } catch (const ss::gate_closed_exception&) {
+        response = cloud_storage::upload_result::cancelled;
+    } catch (const ss::abort_requested_exception&) {
+        response = cloud_storage::upload_result::cancelled;
+    } catch (const std::exception& e) {
+        vlog(_rtclog.error, "failed to upload segment {}: {}", path, e);
+        response = cloud_storage::upload_result::failed;
     }
+    // Cleanup the stream state if upload_segment didn't use it.
+    if (stream_state.has_value()) {
+        co_await stream_state->close();
```

Comment:
The lifetime of the stream_state looks very complicated. The reference is captured in the lambda, then inside the `cloud_storage::remote::upload_segment` the lambda can be invoked and in this case the stream is moved out of the `stream_state`. The `remote` will close it in this case. Otherwise, the `stream_state` is closed here.

The change makes sense though.

Reply:
Agreed that it is complex - at least now it's all in one place. I'm happy to take suggestions to clean this up more broadly (I don't fully understand everything going on here yet).

Reply:
I think it's fine for now but we will refactor this later.